### ISSUE TITLE
False positive with complex nested types

### ISSUE THREAD

**Issue by bthorsted** [microsoft/pyright#10236]
**Describe the bug**
Pyright incorrectly reports that the output type of `data.get(...)` is not assignable to the specified type of the assigned variable.

**Code**

```python
from datetime import date

type ParsedValue = str | int | float | date | None
type CellValue = dict[str, ParsedValue]
type CellValuesList = list[CellValue]
type SheetValues = dict[str, ParsedValue | CellValuesList | CellValue]

def prune_empty(data: SheetValues) -> SheetValues:
    domtrees: list[dict[str, str | int | float | date | None]] = data.get("domtrees", [])
    for domtree in domtrees.copy():
        if not any((domtree["hdspec"], domtree["hdh"], domtree["hdage"])):
            domtrees.remove(domtree)
    smalltrees: list[dict[str, str | int | float | date | None]] = data.get("smalltrees", [])
    for smalltree in smalltrees.copy():
        if not any(
            (smalltree["species"], smalltree["smadia"], smalltree["number"])
        ):
            smalltrees.remove(smalltree)
    return data
```

MWE on [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAJghgFwKYIJYSVdAHMIGyJIBQxCAntpgApwgDOSMAanADYCumAvFPQiCgAfLADsCI4GzCJhhZHIByYUSTKVMAYSRs2rTj1ioAxggDa-EABootBk31cAuuqpRtux0noAZVPyheNn9zDz12Z1dMAGUACyQUL3pAo1MLARs7RhYIzBEwpL8A-J1wgxdiGCRgKGwQDlUAfSQIbAoACngEOAAuKDiEhCSASigAWgA%2BfvjE3Poe4igl2EgBBPmoYP4zGBNzSxtLOVRxOSkZCXk8qGVVJycUrrgAOgBzFHaAIhhVkHXPmxmJzDRbLYB4FYQNaYE6Q6H0Z7GMDYcjtYYLZaYrA1URgAhwUSozq-BJmT6xGD0KjGT5OGw-KF-JBkimxWn0knM8nwd604bo0FYrEM%2BHPP4QMAANyQxMZCRBmPoEHYbHhfS25l2aQOfAEx1OkmkshEXWutyQ90eiBe7wQXyVKvhAKgQIVYIhDt00LEfGVXqZCKRKLRGKF2KguPxhPagrDS3antVTLJVKQxlQ3nZvsdyc%2BDt2cCziehZNEHAgACMkCA%2BbHMQK44q-Un1mKWlKZcWmW6ln8EBwQKJ5HAgA)

**VS Code extension or command-line**
Pylance v2025.3.2

**Comment by erictraut** [comment]
Pyright's behavior here is correct according to type theory and the Python typing spec, so this isn't a bug.

The return type of `data.get` is `ParsedValue | CellValuesList | CellValue`. This type is not assignable to the type `list[dict[str, str | int | float | date | None]]` because of invariance. 

Mypy reports the same error in this case.
